File: core/src/main/java/com/toocol/termio/core/mosh/core/crypto/ByteOrder.java
package com.toocol.termio.core.mosh.core.crypto;

import com.google.common.primitives.Bytes;
import com.google.common.primitives.Longs;
import com.google.common.primitives.Shorts;

import java.nio.ByteBuffer;

/**
 * @author ：JoeZane (joezane.cn@gmail.com)
 * @date: 2022/4/30 18:32
 * @version: 0.0.1
 */
public final class ByteOrder {

    public static java.nio.ByteOrder nativeOrder() {
        return java.nio.ByteOrder.nativeOrder();
    }

    /**
     * whether is the little-endian.
     */
    public static boolean littleEndian() {
        return nativeOrder() == java.nio.ByteOrder.LITTLE_ENDIAN;
    }

    /**
     * transfer little-endian long to big-endian byte[].
     */
    public static byte[] htoBe64(long x) {
        return new byte[]{
                (byte) ((x >> 56) & 0xFF),
                (byte) ((x >> 48) & 0xFF),
                (byte) ((x >> 40) & 0xFF),
                (byte) ((x >> 32) & 0xFF),
                (byte) ((x >> 24) & 0xFF),
                (byte) ((x >> 16) & 0xFF),
                (byte) ((x >> 8) & 0xFF),
                (byte) ((x) & 0xFF)
        };
    }

    /**
     * transfer big-endian bytes to long.
     */
    public static long be64toh(byte[] bytes) {
        return Longs.fromByteArray(bytes);
    }

    /**
     * transfer little-endian short to big-endian byte[].
     */
    public static byte[] htoBe16(short x) {
        return new byte[]{
                (byte) ((x >> 8) & 0xFF),
                (byte) ((x) & 0xFF),
        };
    }

    /**
     * transfer big-endian bytes to short.
     */
    public static short be16toh(byte[] bytes) {
        return Shorts.fromByteArray(bytes);
    }

    public static byte[] bswap64(long x) {
        byte[] array = longBytes(x);
        Bytes.reverse(array);
        return array;
    }

    public static byte[] longBytes(long x) {
        ByteBuffer buffer = ByteBuffer.allocate(8).order(nativeOrder());
        buffer.putLong(0, x);
        return buffer.array();
    }

    public static long toLong(byte[] bytes) {
        ByteBuffer buffer = ByteBuffer.allocate(8).order(nativeOrder());
        buffer.put(0, bytes);
        return buffer.getLong();
    }

    public static byte[] intBytes(int x) {
        ByteBuffer buffer = ByteBuffer.allocate(4).order(nativeOrder());
        buffer.putInt(0, x);
        return buffer.array();
    }

    public static int toInt(byte[] bytes) {
        ByteBuffer buffer = ByteBuffer.allocate(4).order(nativeOrder());
        buffer.put(0, bytes);
        return buffer.getInt();
    }
}

File: utilities/src/util/byte_order.rs
pub struct ByteOrder;

impl ByteOrder {
    /// Whether is the little endian.
    pub fn little_endian() -> bool {
        cfg!(target_endian = "little")
    }

    /// Transfer little-endian long to big-endian [u8].
    pub fn h_to_be_64(x: u64) -> [u8; 8] {
        [
            ((x >> 56) & 0xFF) as u8,
            ((x >> 48) & 0xFF) as u8,
            ((x >> 40) & 0xFF) as u8,
            ((x >> 32) & 0xFF) as u8,
            ((x >> 24) & 0xFF) as u8,
            ((x >> 16) & 0xFF) as u8,
            ((x >> 8) & 0xFF) as u8,
            (x & 0xFF) as u8,
        ]
    }

    /// Transfer big-endian bytes to u64.
    pub fn be_64_to_h(bytes: [u8; 8]) -> u64 {
        u64::from_be_bytes(bytes)
    }

    /// Transfer little-endian u16 to big-endian byte[].
    pub fn h_to_be_16(x: u16) -> [u8; 2] {
        [((x >> 8) & 0xFF) as u8, (x & 0xFF) as u8]
    }

    /// Transfer big-endian bytes to u16.
    pub fn be_16_to_h(bytes: [u8; 2]) -> u16 {
        u16::from_be_bytes(bytes)
    }

    pub fn bswap64(x: u64) -> [u8; 8] {
        let mut array = ByteOrder::long_bytes(x);
        array.reverse();
        array
    }

    pub fn long_bytes(x: u64) -> [u8; 8] {
        if ByteOrder::little_endian() {
            x.to_le_bytes()
        } else {
            x.to_be_bytes()
        }
    }

    pub fn to_long(bytes: [u8; 8]) -> u64 {
        if ByteOrder::little_endian() {
            u64::from_le_bytes(bytes)
        } else {
            u64::from_be_bytes(bytes)
        }
    }

    pub fn int_bytes(x: u32) -> [u8; 4] {
        if ByteOrder::little_endian() {
            x.to_le_bytes()
        } else {
            x.to_be_bytes()
        }
    }

    pub fn to_int(bytes: [u8; 4]) -> u32 {
        if ByteOrder::little_endian() {
            u32::from_le_bytes(bytes)
        } else {
            u32::from_be_bytes(bytes)
        }
    }
}
